ion-schema/src/isl/version_based_isl/mod.rs

```diff
@@ -0,0 +1,396 @@
+//! );
+//!
+//! // create an ISL schema using above IslType
+//! let isl_schema = IslSchema::new("my_schema", vec![], vec![isl_type], vec![], vec![]);
+//!
+//! assert_eq!(isl_schema.types().len(), 1);
+//! assert_eq!(isl_schema.types()[0].name(), &Some("my_type_name".to_owned()));
+//! assert_eq!(isl_schema.types()[0].constraints().len(), 3);
+//! // verify that the last constraint is actually an open content field
+//! assert!(matches!(isl_schema.types()[0].constraints()[2], IslConstraint::Unknown(_)));
+//! ```
+//!
+//! ## Example usage of `version_based_isl` module to create an `IslSchema` using `IslType` for ISL 2.0:
+//! ```
+//! use ion_rs::element::Element;
+//! use ion_schema::isl::version_based_isl::constraints::IslConstraint;
+//! use ion_schema::isl::version_based_isl::{IslSchema, IslTypeRef, IslV2_0, UserReservedFields};
+//! use ion_schema::isl::version_based_isl::constraints::open_content::OpenContent;
+//! use crate::ion_schema::isl::version_based_isl::IslType;
+//! use crate::ion_schema::isl::version_based_isl::constraints::r#type::Type;
+//!
+//! // below code represents an ISL type:
+//! // type:: {
+//! //      name:my_type_name,
+//! //      type: int,
+//! //  }
+//! let isl_type: IslType<IslV2_0> = IslType::named(
+//!     // represents the `name` of the defined type
+//!     "my_type_name".to_owned(),
+//!     vec![
+//!         // represents the `type: int` constraint
+//!         IslConstraint::type_constraint(
+//!             IslTypeRef::named("int")
+//!         ),
+//!     ]
+//! );
+//!
+//! // create an ISL schema using above IslType
+//! let mut  isl_schema = IslSchema::new("my_schema", vec![], vec![isl_type], vec![], vec![]);
+//! // Performing the following operation with ISL 1.0 schema would result in a compile time error.
+//! isl_schema = isl_schema.with_user_reserved_fields(UserReservedFields::new(vec!["foo".to_string(), "bar".to_string()], vec![], vec![]));
+//!
+//! assert_eq!(isl_schema.types().len(), 1);
+//! // verify that the schema has user reserved fields
+//! assert!(isl_schema.user_reserved_fields().is_some());
+//! assert_eq!(isl_schema.user_reserved_fields().unwrap().schema_header_fields().len(), 2);
+//! ```
+use crate::isl::isl_range::Range;
+use crate::isl::isl_type_reference::NullabilityModifier;
+use crate::isl::version_based_isl::constraints::IslConstraint;
+use ion_rs::element::Element;
+use ion_rs::IonType;
+use std::marker::PhantomData;
+
+pub mod constraints;
+
+pub trait IslVersionTrait {}
+
+pub struct IslV1_0 {}
+impl IslVersionTrait for IslV1_0 {}
+
+// TODO: This can be a trait if we have get a new minor version of ISL
+#[derive(Debug, Clone)]
+pub struct IslV2_0 {}
+impl IslVersionTrait for IslV2_0 {}
+
+#[derive(Debug, Clone, Default, PartialEq)]
+pub struct UserReservedFields<V: IslVersionTrait> {
+    schema_header_fields: Vec<String>,
+    schema_footer_fields: Vec<String>,
+    type_fields: Vec<String>,
+    phantom: PhantomData<V>,
+}
+
+impl UserReservedFields<IslV2_0> {
```

Comment:
🗺️  This is only defined for ISL 2.0 and hence would result in a compile time error when user tries to create this with ISL 1.0 schema. Please look at the `IslSchema#with_user_reserved_fields()`.